cpp/src/arrow/compute/exec/filter_node.cc

```diff
@@ -98,12 +106,23 @@ class FilterNode : public ExecNode {
 
   void InputReceived(ExecNode* input, ExecBatch batch) override {
     DCHECK_EQ(input, inputs_[0]);
-
-    auto maybe_filtered = DoFilter(std::move(batch));
-    if (ErrorIfNotOk(maybe_filtered.status())) return;
-
-    maybe_filtered->guarantee = batch.guarantee;
-    outputs_[0]->InputReceived(this, maybe_filtered.MoveValueUnsafe());
+    auto task = [this, batch]() {
+      auto maybe_filtered = DoFilter(std::move(batch));
+      if (ErrorIfNotOk(maybe_filtered.status())) {
+        return maybe_filtered.status();
+      }
+      maybe_filtered->guarantee = batch.guarantee;
+      outputs_[0]->InputReceived(this, maybe_filtered.MoveValueUnsafe());
+      return Status::OK();
+    };
+    auto status = impl_->SubmitTask(task);
+    if (!status.ok()) {
```

Comment:
Why not fold most of this logic into the impl itself?

```cpp
if (!impl_->SubmitTask(task)) {
  StopProducing();
}
```

Reply:
I addressed these comments in the refactored MapNode Class